Add External Tor Usage Documentation

preland opened the pull request:
This PR addresses #1253.

boldsuck commented:
Hi @preland,
external Tor for Haveno Client can be used in 2 ways.

1. Your way with ControlPort: Haveno Client creates a hidden service with jtorctl. You can **not** use HiddenServices options.
Important: Haveno user must be in tor group to use ControlPort.
Add user to tor group on Debian/Ubuntu  `sudo usermod -aG debian-tor <user>`.

2. Or how the seed nodes use the SocksPort. (Haveno user must **not** in tor group)
The HiddenService must be created in /etc/tor/torrc. (You can use all HiddenServices options.)
e.g.:
```
# Haveno incoming anonymity connections
HiddenServiceDir /var/lib/tor/haveno_service/
HiddenServicePort 9999 127.0.0.1:9999
HiddenServicePort 9999 [::1]:9999
```

https://github.com/haveno-dex/haveno/pull/1170#issuecomment-2254246764

Maybe you can link to the seednode documentation ### Install Tor
https://github.com/haveno-dex/haveno/blob/master/docs/deployment-guide.md#install-tor
Install Tor from deb.torproject.org is preferred, Debian backports are also OK. Usually only a few days later. Both are from the same maintainer.

Prove if PoW is enabled: `tor --list-modules`
Tor config e.g. `CookieAuthFile` is preconfigured on Debian derivatives by `/usr/share/tor/tor-service-defaults-torrc`

woodser commented on docs/external-tor-usage.md:
```diff
@@ -0,0 +1,49 @@
+
+### 1. Download and Install Tor
+#### Linux
+You can install Tor through your system's package manager, or directly by going to torproject.org. Note that your package manager's version of tor might be outdated or non-GNU (see below note).
```
Can you give an example using a common package manager? e.g. apt?

preland replied:
All of the issues mentioned should be addressed

woodser commented on docs/external-tor-usage.md:
```diff
@@ -0,0 +1,49 @@
+#### Linux
+You can install Tor through your system's package manager, or directly by going to torproject.org. Note that your package manager's version of tor might be outdated or non-GNU (see below note).
+#### MacOS
+You can install Tor through Homebrew (brew install tor), or directly by going to torproject.org. Note that the Homebrew version of tor might be outdated or non-GNU (see note below).
```
Change to \`brew install tor\` with the backticks, so it's clear it's a command.

woodser commented on docs/external-tor-usage.md:
````diff
@@ -0,0 +1,49 @@
+```
+> The "CookieAuthFile" path should be in a relatively stable and safe place. A good place to put it is in the same directory as torrc.
+
+### The location of your torrc file varies depending on your system:
````
Should be '####' so it's not at same level as main numbers?

woodser replied:
These files do not exist by default after installing (at least on mac). Should they create it manually? Or start tor and stop tor to create the default file?

woodser commented on docs/external-tor-usage.md:
```diff
@@ -0,0 +1,49 @@
+You can install Tor directly by going to torproject.org.
+> For best results, ensure that your version of tor matches the one currently used by Haveno. To improve loading times, it is highly recommended to use the GNU build, as it has support for POW. You can check if your build is GNU by running tor --version and looking for a response mentioning GNU.
+
+### 2. Add the following lines to your torrc:
```
Please change to "torcc file:" for clarity / consistency.

woodser commented on docs/external-tor-usage.md:
````diff
@@ -0,0 +1,49 @@
+CookieAuthentication 1
+CookieAuthFile <path to cookie auth file>
+```
+> The "CookieAuthFile" path should be in a relatively stable and safe place. A good place to put it is in the same directory as torrc.
````
Maybe show an example path to use for CookieAuthFile?

Something like:

"For example, on Linux you could use `CookieAuthFile /etc/tor/torrc`"

woodser commented on docs/external-tor-usage.md:
````diff
@@ -0,0 +1,49 @@
+#### Windows
+```\Browser\TorBrowser\Data\Tor\torrc``` (when using the Tor Browser Bundle)
+
+> The above locations are best guesses, and may be incorrect depending on your installation method or platform distribution. In the case that you cannot find the correct directory to place your torrc, tor can be ran in the following format to use a custom torrc: ```tor -f <torrc>```
````
Suggested tweaks:

"The above locations may vary depending on your installation method or platform distribution. If you cannot find the correct directory to place your torrc file, you can specify a custom path to torcc in the following format: ```tor -f <path to torcc```

woodser commented on docs/external-tor-usage.md:
````diff
@@ -0,0 +1,49 @@
+## Using External Tor with Haveno:
+Run the Haveno binary with the following extra flags:
+```
+--torControlPort=9051 --torControlCookieFile=<path to cookie auth file> --torControlUseSafeCookieAuth=true
````
Just confirming, the user must repeat these definitions, both in the torcc file and when invoking the Haveno binary?

preland replied:
Yes; this tells Haveno to use the tor port and authentication for the external tor instead of the internal tor

woodser commented on docs/external-tor-usage.md:
````diff
@@ -0,0 +1,49 @@
+```
+> Replace the ```<path to cookie auth file>``` with the location you specified in step 2.
+
+> Depending on how your version of Haveno was distributed, you may not have direct access to the haveno binary. In this case, you should look inside your Haveno distribution for the binary, and then run the binary using the above flags. For example, on MacOS the binary will be located within the app at Haveno.app/Contents/MacOS/Haveno, or something similar.
````
"haveno" should be capitalized for consistency.

woodser commented:
I can't get tor to create my cookie authentication file for some reason. Not sure what I'm doing wrong, but it's not created when tor is started. I've configured my torrc file (installed through homebrew):

```
ControlPort 9051
CookieAuthentication 1
CookieAuthFile /opt/homebrew/etc/tor/control_auth_cookie
```

Tor has permission to write to that directory.

preland commented:
I have made a few changes to the instructions which should address some issues with the original writing:

-Removed mentioning of using package managers for downloading Tor. This added extra complexity and confusion to the document, as well as went against the recommendations of the Tor Project.
-replaced the vagueness of the CookieAuthFile name. While the file name is technically arbitrary, I doubt many will lose sleep over naming the file “control_auth_cookie”.
-Added some extra context for killing tor (most likely unnecessary, but should be mentioned just in case)
-Added running tor with the ```-f``` flag in step 4.

boldsuck commented:
Hi @preland
I just saw this by chance https://github.com/bisq-network/bisq/pull/1935
I think Haveno | Bisq supports Tor ControlPassword or SafeCookieAuth and not CookieAuth.
[Tor Authentication via Control Port ](https://gist.github.com/ndsamuelson/3c83f38ae470c82ff87d2653af11716b)

I have only used Tor ControlPassword with Netlayer / jtorctl as described by Whonix Dev Patrick.
This eliminates all problems with read/write permissions of the control_auth_cookie file. (Cookie is created anew every time tor restarts.)

With every restart Haveno writes a new:
`~/.local/share/Haveno/xmr_stagenet/tor/torrc`
A permanent Tor configuration, in addition to Haveno standard torrc-entries, can be easily created using the tor `%include` config option, e.g.: `./torrc.local`:
`--torrcOptions=%include ~/.local/share/Haveno/xmr_stagenet/tor/torrc.local` cmdline or in haveno.properties

preland commented:
Do you think that ControlPassword would be a better fit than SafeCookieAuth for our use case?

boldsuck commented:
It's not better, I just wanted to point out the alternative. There seem to be problems with SafeCookieAuth on some systems bc. read/write permissions of the control_auth_cookie file. If your steps work so far, @woodser should merge them.¹ I don't have Whonix. [Is it possible that Whonix has `notrequied` as the default torControlPassword?](https://www.whonix.org/wiki/Bisq#Usage)

¹I will then add to the docu how to configure Haveno Client with externalTor without using jtorctl. @woodser had @fa2a5qj3 change the code for the client as well. Unfortunately, this option is rarely used and all instructions refer to use system tor the old jtorctl/Netlayer bisq1 way.

woodser commented:
Documentation to run with external tor is merged here: https://github.com/haveno-dex/haveno/pull/1627
